### ISSUE TITLE
RGBWW support

### ISSUE THREAD

**Issue by vandalon** [basnijholt/adaptive-lighting#289]
Hi Bas,

First of all thanks for the great plugin, i've been using it for a while now with my zigbee lights and all is great!
But now i've added some RGBWW lights to my setup and those don't seem to work with adaptive lighting. I glanced over the code and I don't see any color_temperature_to_rgbww anywhere so i'm guessing this is just missing :) Since RGBWW light don't support CT I think it would be great if you can add RGBWW color mode support.

Thanks for all the great stuff!
Joris

**Comment by RubenKelevra** [comment]
RGBWW means RGB plus Warm-White, right? 

Can you enable Warm-White output and mix it with regular RGB?

**Comment by vandalon** [comment]
@RubenKelevra No, RGB warm + cold white, so 2 white channels instead of one. I know it's also called RGBCCT but in HA this is called RGBWW

**Comment by RubenKelevra** [comment]
@vandalon I got like 20 lamps of this type, they work fine with Adaptive Lighting. Can you elaborate what's not working for you?

**Comment by vandalon** [comment]
@RubenKelevra Are you sure, are they also running with color_mode: rgbww in home assistant?

**Comment by RubenKelevra** [comment]
Ah yeah, just realised the same thing. They do run with HS/RGB/XY or color temperature.

So your lights *only* support RGBWW?

**Comment by vandalon** [comment]
Well i'm running esphome on the controller, but RGBWW would be the most flexible option. Mostly because you can combine that white channels and RGB channels in a very flexible way. When using RGBWW there is no CT mode.

**Comment by RubenKelevra** [comment]
@vandalon ah! Gotcha. Yeah using RGB to "emulate" real warm/cool white LEDs is quite bad. I assumed that your lights would convert the CT-mode commands to just the white LEDs and the RGB/HSV commands to the colored LEDs – as that's how my Zigbee LED strips works, which is also RGBWW.

Yeah, will have a look at this on this goal. Should be pretty easy.

- [Dynamically switch between RGB/XY and White-Temperature mode to allow very cold and very warm tones outside the range of the White-Mode of lights](https://github.com/basnijholt/adaptive-lighting/discussions/291#discussioncomment-2886270)

I'll close the ticket, as it's now on the roadmap :)

**Comment by th3w1zard1** [comment]
I've been looking into this issue, and it seems to be inaccurate on hassio itself?
https://github.com/home-assistant/core/issues/54216
Are you able to get accurate RGBWW colors using `light.turn_on` inside of HASS? Are you certain it's just `adaptive-lighting` that's inaccurate?

**Comment by th3w1zard1** [comment]
@vandalon This should already be in adaptive lighting as its already in home assistant. If your light doesn't support the right color output, home assistant will convert to whatever your bulb supports inside its own code.
see `homeassistant\components\light\light.py`

**Comment by th3w1zard1** [comment]
I believe I've found a function that will convert ct to rgbcct for more accurate values:

def kelvin_to_rgbcct(kelvin):
    """
    Convert a color temperature in Kelvin to RGBCCT values.

    :param kelvin: The color temperature in Kelvin.
    :type kelvin: int or float
    :return: A tuple of five values representing the RGBCCT values.
    :rtype: tuple
    """
    # Convert the color temperature to mireds
    mireds = 1000000 / kelvin

    # Calculate the CT value (CCT range is typically 153 to 500)
    ct = 449 * mireds ** 3 - 3525 * mireds ** 2 + 6823.3 * mireds + 5520.33

    # Calculate the red, green, and blue values
    if ct <= 153:
        red = 255
        green = (ct * 255) // 153
        blue = 0
    elif ct <= 255:
        red = (255 * (255 - (ct - 153) * 5)) // 255
        green = 255
        blue = 0
    elif ct <= 400:
        red = 0
        green = 255
        blue = ((ct - 255) * 255) // (400 - 255)
    elif ct <= 500:
        red = 0
        green = (255 * (500 - ct)) // (500 - 400)
        blue = 255
    else:
        red = 0
        green = 0
        blue = 255

    # Calculate the CCT value
    cct = 1000000 / mireds

    # Return the RGBCCT values as a tuple
    return red, green, blue, 255, int(cct)

**Comment by basnijholt** [comment]
Related: https://github.com/basnijholt/adaptive-lighting/issues/118

**Comment by basnijholt** [comment]
@th3w1zard1, do you remember where you found that (https://github.com/basnijholt/adaptive-lighting/issues/289#issuecomment-1504843965)?

**Comment by th3w1zard1** [comment]
I don't, I might have just wrote it using wikipedia to help me with the conversions.

**Comment by lloeki** [comment]
I just set up GU10 and E27 Shelly RGB bulbs (moved from IKEA Tradfri white-only bulbs which were working perfectly in that regard). I appear to be experiencing #118.

> I've been looking into this issue, and it seems to be inaccurate on hassio itself?
https://github.com/home-assistant/core/issues/54216

This has been merged and released.

> Are you able to get accurate RGBWW colors using light.turn_on inside of HASS?

I am. Brightness slider works, RGB colour picking works, white temperature slider works, all completely independently. I don't know if this uses `light.turn_on` behind the scenes.

> Are you certain it's just adaptive-lighting that's inaccurate?

Given the above, yes. With adaptive-lighting, colour and brightness were coupled in some way, e.g pure red was super dim, some other colours were much righters - can't recall which exactly -, white was super bright and of fixed temperature (interestingly enough temperature wasn't entirely consistent across all bulbs), in all cases brightness slider was ineffective.

I had to fully disable the adaptive-lighting integration to be able to debug this (when the component was enabled adaptive-lighting kept on mucking with the lights in various ways, which made debugging inconsistent and annoying), which made lights *immediately* work properly with HASS. Weirdly enough, adaptive-lighting seemed to have an influence even with all light control toggles off, and also on lights removed from the light control list.

**Comment by lloeki** [comment]
May I ask, what's the semantic of the hardcoded `255` here?

```
def kelvin_to_rgbcct(kelvin):
  ...
  # Return the RGBCCT values as a tuple
  return red, green, blue, 255, int(cct)
```

**Comment by th3w1zard1** [comment]
I am not sure, sorry. Been almost a year

**Comment by lloeki** [comment]
Totally fair, thanks for taking the time to answer!

**Comment by basnijholt** [comment]
[_Copy pasting this message in a few recent open issues_]

I just wanted to take a moment to express my heartfelt thanks to everyone that is active in this repo. Your contributions, from answering questions to addressing issues, have been invaluable. It's amazing to see how supportive and helpful our community is!

Adaptive Lighting is all about enhancing your living spaces with smart, sunlight-responsive lighting. We've had quite a few discussions and open issues recently, and I see this as a positive sign of our community's engagement and growth. If you come across anything in the documentation that's unclear or if you have suggestions for improvement, _**please don't hesitate to share!**_. Your feedback is crucial for making Adaptive Lighting better for everyone.

On a personal note, I've recently welcomed twin boys into my family, which has been an incredible and life-changing experience. As you can imagine, my time is now more limited, and while I'm doing my best to keep up with the project, there may be delays in my responses. I appreciate your understanding and patience during this time.

Rest assured, I'm fully committed to addressing any bugs, especially those related to new Home Assistant updates, as swiftly as possible. I understand that many issues may stem from hardware limitations or misunderstandings about things like Zigbee groups. Your continued support and collaboration in helping each other out not only strengthen our community but also enhance the Adaptive Lighting experience for all.

Thank you once again for your understanding, patience, and support. Let's keep our houses well lit and adaptive for maximal enjoyment of life! 🌞🏠🌙
